graph/src/data/subgraph/schema.rs

```diff
@@ -591,6 +616,80 @@ impl From<super::MappingEventHandler> for EthereumContractEventHandlerEntity {
     }
 }
 
+#[derive(Debug)]
+struct EthereumContractDataSourceTemplateEntity {
+    kind: String,
+    network: Option<String>,
+    name: String,
+    source: EthereumContractDataSourceTemplateSourceEntity,
+    mapping: EthereumContractMappingEntity,
+}
+
+impl TypedEntity for EthereumContractDataSourceTemplateEntity {
+    const TYPENAME: &'static str = "EthereumContractDataSourceTemplate";
+    type IdType = String;
+}
+
+impl EthereumContractDataSourceTemplateEntity {
+    fn write_operations(self, id: &str) -> Vec<EntityOperation> {
+        let mut ops = vec![];
+
+        let source_id = format!("{}-source", id);
+        ops.extend(self.source.write_operations(&source_id));
+
+        let mapping_id = format!("{}-mapping", id);
+        ops.extend(self.mapping.write_operations(&mapping_id));
+
+        let mut entity = Entity::new();
+        entity.set("id", id);
+        entity.set("kind", self.kind);
+        entity.set("network", self.network);
+        entity.set("name", self.name);
+        entity.set("source", source_id);
+        entity.set("mapping", mapping_id);
+        ops.push(set_entity_operation(Self::TYPENAME, id, entity));
+
+        ops
+    }
+}
+
+impl From<&super::DataSourceTemplate> for EthereumContractDataSourceTemplateEntity {
+    fn from(template: &super::DataSourceTemplate) -> Self {
+        Self {
+            kind: template.kind.clone(),
+            name: template.name.clone(),
+            network: template.network.clone(),
+            source: template.source.clone().into(),
+            mapping: EthereumContractMappingEntity::from(&template.mapping),
+        }
+    }
+}
+
+#[derive(Clone, Debug, Hash, Eq, PartialEq)]
+struct EthereumContractDataSourceTemplateSourceEntity {
+    abi: String,
+}
+
+impl TypedEntity for EthereumContractDataSourceTemplateSourceEntity {
```

Comment:
Echoing what I said in Discord earlier, I had to read this chunk three times to convince myself that it was about exactly two new entities, `EthereumContractDataSourceTemplateSourceEntity` and  `EthereumContractDataSourceTemplateEntity`. It's incredibly hard to keep the two apart, and we should figure out some naming scheme that makes these names a little easier to read and tell apart. Maybe something like `EthCDSTemplateEntity` and `EthCDSTemplateSourceEntity`; I am not in love with these either, but at least I can now see easily that they are different length.